connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteGrpcResponseSender.scala

```diff
@@ -46,72 +57,118 @@ private[connect] class ExecuteGrpcResponseSender[T](grpcObserver: StreamObserver
 
   /**
    * Attach to the executionObserver, consume responses from it, and send them to grpcObserver.
+   *
+   * In non reattachable execution, it will keep sending responses until the query finishes. In
+   * reattachable execution, it can finish earlier after reaching a time deadline or size limit.
+   *
+   * After this function finishes, the grpcObserver is closed with either onCompleted or onError.
+   *
    * @param lastConsumedStreamIndex
    *   the last index that was already consumed and sent. This sender will start from index after
    *   that. 0 means start from beginning (since first response has index 1)
-   *
-   * @return
-   *   true if the execution was detached before stream completed. The caller needs to finish the
-   *   grpcObserver stream false if stream was finished. In this case, grpcObserver stream is
-   *   already completed.
    */
-  def run(
-      executionObserver: ExecuteResponseObserver[T],
-      lastConsumedStreamIndex: Long): Boolean = {
+  def run(lastConsumedStreamIndex: Long): Unit = {
+    logDebug(
+      s"GrpcResponseSender run for $executeHolder, " +
+        s"reattachable=${executeHolder.reattachable}, " +
+        s"lastConsumedStreamIndex=$lastConsumedStreamIndex")
+
     // register to be notified about available responses.
     executionObserver.attachConsumer(this)
 
     var nextIndex = lastConsumedStreamIndex + 1
     var finished = false
 
+    // Time at which this sender should finish if the response stream is not finished by then.
+    val deadlineTimeMillis = if (!executeHolder.reattachable) {
+      Long.MaxValue
+    } else {
+      val confSize =
+        SparkEnv.get.conf.get(CONNECT_EXECUTE_REATTACHABLE_SENDER_MAX_STREAM_DURATION).toLong
+      if (confSize > 0) System.currentTimeMillis() + 1000 * confSize else Long.MaxValue
```

Comment:
Does this not return a duration or something similar to avoid the manual conversion?